### PR TITLE
HHH-16826 - IN-Clause Parameter Padding should grow exponentially for Dialects with InExpressionCountLimit

### PR DESCRIPTION
**Pull request by Adrodoc**
See [HHH-16826](https://hibernate.atlassian.net/browse/HHH-16826).

In the first commit I added some parameterized tests to `MaxInExpressionParameterPaddingTest` since they are much less verbose for this kind of test. I also added some of the constellation of the existing tests to the parameterized test data. Now some constellations are tested twice, since I did not dare to remove the existing tests. Some of the existing tests have annotations like `@TestForIssue(jiraKey = "HHH-14109")`. If I were to remove the existing tests in favor of the parameterized tests I wasn't sure what to do with those annotations. Since the annotation is not repeatable I can't add multiple annotations to the parameterized test and I don't know what you use those annotations for.

To avoid the test duplication I removed the duplicated old tests in a second commit and added comments to the parameterized test data to point to the relevant Jira tickets. If this approach is fine I would suggest to squash the two commits into one. If this is not fine I would suggest to drop the second commit.

[HHH-16826]: https://hibernate.atlassian.net/browse/HHH-16826?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 1]
```diff
@@ -6912,69 +6886,58 @@ else if ( bindValueCount < bindValuePaddingCount ) {
...
-				itemNumber = 0;
-				while ( iterator.hasNext() && itemNumber < inExprLimit ) {
-					final Expression listExpression = iterator.next();
-					appendSql( separator );
-					itemAccessor.apply( listExpression ).accept( this );
-					separator = COMMA_SEPARATOR;
-					itemNumber++;
-				}
-			} while ( iterator.hasNext() );
+	private void appendInClauseSeperator(InListPredicate inListPredicate) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:6921]:
```suggestion
	private void appendInClauseSeparator(InListPredicate inListPredicate) {
```

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 2]
```diff
@@ -6912,69 +6886,58 @@ else if ( bindValueCount < bindValuePaddingCount ) {
 		String separator = NO_SEPARATOR;
 
 		final Iterator<Expression> iterator = listExpressions.iterator();
-		int itemNumber = 0;
-		while ( iterator.hasNext() && ( inExprLimit == 0 || itemNumber < inExprLimit ) ) {
-			final Expression listExpression = itemAccessor.apply( iterator.next() );
+		Expression listExpression = null;
+		for ( int i = 0; i < bindValueCountWithPadding; i++ ) {
+			if ( inExprLimit > 0 && i % inExprLimit == 0 && i != 0 ) {
+				appendInClauseSeperator( inListPredicate );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:6892]: Here a simple rewrite that allows avoiding the costly modulo operation.

```suggestion
		Expression listExpression = null;
		int clauseItemNumber = 0;
		for ( int i = 0; i < bindValueCountWithPadding; i++, clauseItemNumber++ ) {
			if ( inExprLimit > 0 && inExprLimit == clauseItemNumber ) {
				clauseItemNumber = 0;
				appendInClauseSeperator( inListPredicate );
```

**Review by beikov — COMMENTED** [review]
I like it! Sorry it took so long for me to review this. With some small adjustments I'd consider merging it :)

**Comment by Adrodoc** [comment]
> I like it!

@beikov I assume this also means you like the parameterized tests, so I will squash the two commits into one.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java (hunk 2 above)
Comment (Adrodoc) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: Looks great. I just took this as is.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java (hunk 1 above)
Comment (Adrodoc) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: I squashed the fix for this typo into the first commit.

**Comment by Adrodoc** [comment]
@beikov Thanks for the review 👍
I applied both of your fixes pretty much as is (for the first change the caller of the method also had to be adjusted).
Feel free to merge the PR anytime :)

**Review by beikov — APPROVED** [review]

**Comment by beikov** [comment]
Thanks for your great work! Looking forward to your next contribution ;)

**Pull request merged**